File: 0426-convert-binary-search-tree-to-sorted-doubly-linked-list/0426-convert-binary-search-tree-to-sorted-doubly-linked-list.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;

    Node() {}

    Node(int _val) {
        val = _val;
        left = NULL;
        right = NULL;
    }

    Node(int _val, Node* _left, Node* _right) {
        val = _val;
        left = _left;
        right = _right;
    }
};
*/

class Solution {
public:
    Node* prev = nullptr;
    Node* first = nullptr;
    Node* last = nullptr;
    
    void traverse(Node* node) {
        if(!node) return;
        traverse(node->left);
        if(!prev) first = node;
        
        if(prev) {
            prev->right = node;
            node->left = prev;
        }
        if(!node->right) last = node;
        prev = node;
        traverse(node->right);
    }
    Node* treeToDoublyList(Node* root) {
        if(!root) return nullptr;
        traverse(root);
        first->left = last;
        last->right = first;
        return first;
    }
};

File: 0426-convert-binary-search-tree-to-sorted-doubly-linked-list/0426-convert-binary-search-tree-to-sorted-doubly-linked-list.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val,Node _left,Node _right) {
        val = _val;
        left = _left;
        right = _right;
    }
};
*/

class Solution {
    Node prev = null;
    Node first = null;
    Node last = null;
    public void inorderTraversal(Node node) {
        //Base Case
        if(node == null) return;
        
        //left and right recursive call for DFS
        inorderTraversal(node.left);
        //INORDER AREA ->>>>>>>>>
        //save first node
        if(prev == null) first = node;
        //add links in between nodes
        if(prev != null) {
            prev.right = node;
            node.left = prev;
        }
        //save last node
        if(node.right == null) last = node;
        //update our prev node
        prev = node;
        //INORDER AREA ->>>>>>>>>
        inorderTraversal(node.right);
    }
    public Node treeToDoublyList(Node root) {
        //edge case
        if(root==null) return null;
        inorderTraversal(root);
        // complete 2 links of first and last nodes
        first.left = last;
        last.right = first;
        
        return first;
    }
}